### DISCUSSION TITLE
TFLite support

### DISCUSSION THREAD

**Discussion by costa-victor** [Moddable-OpenSource/moddable#1290] [Ideas]
Hello all,

I've been playing with TFLite for ESP32 using the ESP-IDF recently and I was interested to contribute and bring this to Moddable environment as I believe it would be an interesting feature to have.

Any ideas or suggestions are very welcome, thank you in advance!

## Approach 1
This is the approach I have in mind which I believe it would work even though it seems to require some good effort to make it happen.

1. Take the [ESP-TFLite component](https://github.com/espressif/esp-tflite-micro)
2. Create an [JS interface](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20in%20C.md#javascript--language-syntax-extension) to bind the C functions into JS

## Approach 2
I'm not sure how much functions I will have to integrate into this new JS module to bring the TFLite to Moddable, because of that I thought in another approach which I'm not sure if it would work.

I found a [TFLite library in JS already](https://js.tensorflow.org/api_tflite/0.0.1-alpha.4/), therefore the interface with the ESP-TFLite which is in C wouldn't be required. Do you think it would work?

Thank you in advance and best regards.

**Comment by phoddie** [comment]
>  was interested to contribute and bring this to Moddable environment as I believe it would be an interesting feature to have.

Agreed!

"Approach 1" seems the most practical.

> I found a [TFLite library in JS already](https://js.tensorflow.org/api_tflite/0.0.1-alpha.4/), therefore the interface with the ESP-TFLite which is in C wouldn't be required. Do you think it would work?

Probably not. The documentation of that notes:

> This library is a wrapper of TFLite interpreter. It is packaged in a WebAssembly binary that runs in a browser. 

The Moddable SDK doesn't provide WebAssembly and certainly isn't a browser. ;) Even if it did run, it seems like it could be quite a bit slower than running natively.
